`RuntimeException: Failed to create Kubernetes pod` /  `KubernetesClientException: An error has occurred.` / `SocketTimeoutException: connect timed out`

Another error, this time using less threads (20):
```
[error] [ForkJoinPool-2-worker-39] INFO com.spotify.hype.Submitter - Submitting gs://hype-test/manifest-2imva9o4.txt to RunEnvironment{yamlPath=Optional.empty, secretMounts=[Secret{name=gcp-key, mountPath=/etc/gcloud}], volumeMounts=[VolumeMount{volume=PersistentDisk{pdName=ml-share}, mountPath=/usr/share/volume, readOnly=true}], resourceRequests={cpu=21}}
[error] [ForkJoinPool-2-worker-39] INFO com.spotify.hype.runner.DockerRunner - Mounting ml-share readOnly at /usr/share/volume
[error] Exception in thread "main" java.lang.RuntimeException: Failed to create Kubernetes pod
[error] 	at com.spotify.hype.runner.KubernetesDockerRunner.run(KubernetesDockerRunner.java:101)
[error] 	at com.spotify.hype.Submitter.runOnCluster(Submitter.java:147)
[error] 	at com.spotify.hype.HypeSubmitter$class.submit(HypeSubmitter.scala:10)
[error] 	at com.spotify.hype.GkeSubmitter.submit(GkeSubmitter.scala:4)
[error] 	at com.spotify.hyperflo.examples.CrossValInternal$$anonfun$3.apply(CrossValInternal.scala:166)
[error] 	at com.spotify.hyperflo.examples.CrossValInternal$$anonfun$3.apply(CrossValInternal.scala:165)
[error] 	at scala.collection.parallel.AugmentedIterableIterator$class.map2combiner(RemainsIterator.scala:115)
[error] 	at scala.collection.parallel.immutable.ParVector$ParVectorIterator.map2combiner(ParVector.scala:62)
[error] 	at scala.collection.parallel.ParIterableLike$Map.leaf(ParIterableLike.scala:1054)
[error] 	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply$mcV$sp(Tasks.scala:49)
[error] 	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:48)
[error] 	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:48)
[error] 	at scala.collection.parallel.Task$class.tryLeaf(Tasks.scala:51)
[error] 	at scala.collection.parallel.ParIterableLike$Map.tryLeaf(ParIterableLike.scala:1051)
[error] 	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.internal(Tasks.scala:159)
[error] 	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.internal(Tasks.scala:443)
[error] 	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.compute(Tasks.scala:149)
[error] 	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.compute(Tasks.scala:443)
[error] 	at scala.concurrent.forkjoin.RecursiveAction.exec(RecursiveAction.java:160)
[error] 	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
[error] 	at scala.concurrent.forkjoin.ForkJoinTask.doJoin(ForkJoinTask.java:341)
[error] 	at scala.concurrent.forkjoin.ForkJoinTask.join(ForkJoinTask.java:673)
[error] 	at scala.collection.parallel.ForkJoinTasks$WrappedTask$class.sync(Tasks.scala:378)
[error] 	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.sync(Tasks.scala:443)
[error] 	at scala.collection.parallel.ForkJoinTasks$class.executeAndWaitResult(Tasks.scala:426)
[error] 	at scala.collection.parallel.ForkJoinTaskSupport.executeAndWaitResult(TaskSupport.scala:56)
[error] 	at scala.collection.parallel.ParIterableLike$ResultMapping.leaf(ParIterableLike.scala:958)
[error] 	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply$mcV$sp(Tasks.scala:49)
[error] 	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:48)
[error] 	at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:48)
[error] 	at scala.collection.parallel.Task$class.tryLeaf(Tasks.scala:51)
[error] 	at scala.collection.parallel.ParIterableLike$ResultMapping.tryLeaf(ParIterableLike.scala:953)
[error] 	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.compute(Tasks.scala:152)
[error] 	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.compute(Tasks.scala:443)
[error] 	at scala.concurrent.forkjoin.RecursiveAction.exec(RecursiveAction.java:160)
[error] 	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
[error] 	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
[error] 	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
[error] 	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
[error] Caused by: io.fabric8.kubernetes.client.KubernetesClientException: An error has occurred.
[error] 	at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:61)
[error] 	at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:52)
[error] 	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.get(BaseOperation.java:201)
[error] 	at com.spotify.hype.runner.KubernetesDockerRunner.blockUntilComplete(KubernetesDockerRunner.java:114)
[error] 	at com.spotify.hype.runner.KubernetesDockerRunner.run(KubernetesDockerRunner.java:97)
[error] 	... 38 more
[error] Caused by: java.net.SocketTimeoutException: connect timed out
[error] 	at java.net.PlainSocketImpl.socketConnect(Native Method)
[error] 	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
[error] 	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
[error] 	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
[error] 	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
[error] 	at java.net.Socket.connect(Socket.java:589)
[error] 	at okhttp3.internal.platform.Platform.connectSocket(Platform.java:124)
[error] 	at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.java:220)
[error] 	at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:146)
[error] 	at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:186)
[error] 	at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:121)
[error] 	at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:100)
[error] 	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42)
[error] 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
[error] 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
[error] 	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
[error] 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
[error] 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
[error] 	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
[error] 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
[error] 	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:120)
[error] 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
[error] 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
[error] 	at io.fabric8.kubernetes.client.utils.HttpClientUtils$2.intercept(HttpClientUtils.java:93)
[error] 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
[error] 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
[error] 	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:179)
[error] 	at okhttp3.RealCall.execute(RealCall.java:63)
[error] 	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:239)
[error] 	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:234)
[error] 	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleGet(OperationSupport.java:230)
[error] 	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.handleGet(BaseOperation.java:745)
[error] 	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.get(BaseOperation.java:194)
[error] 	... 40 more
java.lang.RuntimeException: Nonzero exit code returned from runner: 1
	at scala.sys.package$.error(package.scala:27)
```

FYI just got this fail several times in a row in our build. looks like the same thing.

```INFO: Exception in thread "main" java.lang.RuntimeException: Failed to create Kubernetes pod
INFO: 	at com.spotify.hype.runner.KubernetesDockerRunner.run(KubernetesDockerRunner.java:100)
INFO: 	at com.spotify.hype.Submitter.runOnCluster(Submitter.java:147)
INFO: 	at com.spotify.hype.HypeSubmitter$class.submit(HypeSubmitter.scala:10)
INFO: 	at com.spotify.hype.GkeSubmitter.submit(GkeSubmitter.scala:4)
INFO: 	at com.spotify.ads.adserver.pipelines.hype.HypeRunner$.main(HypeRunner.scala:27)
INFO: 	at com.spotify.ads.adserver.pipelines.hype.HypeRunner.main(HypeRunner.scala)
INFO: Caused by: io.fabric8.kubernetes.client.KubernetesClientException: An error has occurred.
INFO: 	at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:61)
INFO: 	at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:52)
INFO: 	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.get(BaseOperation.java:201)
INFO: 	at com.spotify.hype.runner.KubernetesDockerRunner.blockUntilComplete(KubernetesDockerRunner.java:113)
INFO: 	at com.spotify.hype.runner.KubernetesDockerRunner.run(KubernetesDockerRunner.java:96)
INFO: 	... 5 more
INFO: Caused by: java.net.SocketTimeoutException: timeout
INFO: 	at okio.Okio$4.newTimeoutException(Okio.java:230)
INFO: 	at okio.AsyncTimeout.exit(AsyncTimeout.java:285)
INFO: 	at okio.AsyncTimeout$2.read(AsyncTimeout.java:241)
INFO: 	at okio.RealBufferedSource.indexOf(RealBufferedSource.java:345)
INFO: 	at okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:217)
INFO: 	at okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:211)
INFO: 	at okhttp3.internal.http1.Http1Codec.readResponseHeaders(Http1Codec.java:189)
INFO: 	at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:67)
INFO: 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
INFO: 	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45)
INFO: 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
INFO: 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
INFO: 	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
INFO: 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
INFO: 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
INFO: 	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
INFO: 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
INFO: 	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:120)
INFO: 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
INFO: 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
INFO: 	at io.fabric8.kubernetes.client.utils.HttpClientUtils$2.intercept(HttpClientUtils.java:95)
INFO: 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
INFO: 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
INFO: 	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:179)
INFO: 	at okhttp3.RealCall.execute(RealCall.java:63)
INFO: 	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:239)
INFO: 	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:234)
INFO: 	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleGet(OperationSupport.java:230)
INFO: 	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.handleGet(BaseOperation.java:745)
INFO: 	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.get(BaseOperation.java:194)
INFO: 	... 7 more
INFO: Caused by: java.net.SocketException: Socket closed
INFO: 	at java.net.SocketInputStream.read(SocketInputStream.java:204)
INFO: 	at java.net.SocketInputStream.read(SocketInputStream.java:141)
INFO: 	at sun.security.ssl.InputRecord.readFully(InputRecord.java:465)
INFO: 	at sun.security.ssl.InputRecord.read(InputRecord.java:503)
INFO: 	at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:983)
INFO: 	at sun.security.ssl.SSLSocketImpl.readDataRecord(SSLSocketImpl.java:940)
INFO: 	at sun.security.ssl.AppInputStream.read(AppInputStream.java:105)
INFO: 	at okio.Okio$2.read(Okio.java:139)
INFO: 	at okio.AsyncTimeout$2.read(AsyncTimeout.java:237)
INFO: 	... 34 more```

--- a/hype-common/pom.xml
+++ b/hype-common/pom.xml
@@ -11,6 +11,11 @@
 
   <dependencies>
     <dependency>
+      <groupId>com.google.cloud</groupId>
+      <artifactId>google-cloud-storage</artifactId>
+    </dependency>
+
+    <dependency>
       <groupId>com.esotericsoftware</groupId>
       <artifactId>kryo-shaded</artifactId>
       <version>4.0.0</version>

--- a/hype-common/src/main/java/com/spotify/hype/FluentBackoff.java
+++ b/hype-common/src/main/java/com/spotify/hype/FluentBackoff.java
@@ -16,7 +16,7 @@
  * -/-/-
  */
 
-package com.spotify.hype.gcs;
+package com.spotify.hype;
 
 import static com.google.common.base.Preconditions.checkArgument;
 
@@ -132,7 +132,7 @@ public final class FluentBackoff {
    */
   public FluentBackoff withMaxCumulativeBackoff(Duration maxCumulativeBackoff) {
     checkArgument(maxCumulativeBackoff.isLongerThan(Duration.ZERO),
-        "maxCumulativeBackoff %s must be at least 1 millisecond", maxCumulativeBackoff);
+                                "maxCumulativeBackoff %s must be at least 1 millisecond", maxCumulativeBackoff);
     return new FluentBackoff(
         exponent, initialBackoff, maxBackoff, maxCumulativeBackoff, maxRetries);
   }

--- a/hype-common/src/main/java/com/spotify/hype/FluentBackoff.java
+++ b/hype-common/src/main/java/com/spotify/hype/FluentBackoff.java
@@ -132,7 +132,7 @@ public final class FluentBackoff {
    */
   public FluentBackoff withMaxCumulativeBackoff(Duration maxCumulativeBackoff) {
     checkArgument(maxCumulativeBackoff.isLongerThan(Duration.ZERO),
-                                "maxCumulativeBackoff %s must be at least 1 millisecond", maxCumulativeBackoff);
+        "maxCumulativeBackoff %s must be at least 1 millisecond", maxCumulativeBackoff);
     return new FluentBackoff(
         exponent, initialBackoff, maxBackoff, maxCumulativeBackoff, maxRetries);
   }

--- a/hype-gcs/pom.xml
+++ b/hype-gcs/pom.xml
@@ -11,6 +11,12 @@
 
   <dependencies>
     <dependency>
+      <groupId>com.spotify</groupId>
+      <artifactId>hype-common</artifactId>
+      <version>0.0.18-SNAPSHOT</version>
+    </dependency>
+
+    <dependency>
       <groupId>com.google.cloud</groupId>
       <artifactId>google-cloud-storage</artifactId>
     </dependency>

--- a/hype-gcs/src/main/java/com/spotify/hype/gcs/StagingUtil.java
+++ b/hype-gcs/src/main/java/com/spotify/hype/gcs/StagingUtil.java
@@ -18,6 +18,7 @@
 
 package com.spotify.hype.gcs;
 
+
 import static com.google.cloud.storage.contrib.nio.CloudStorageOptions.withMimeType;
 import static java.nio.file.StandardOpenOption.CREATE_NEW;
 import static java.nio.file.StandardOpenOption.WRITE;
@@ -37,6 +38,7 @@ import com.google.common.io.Files;
 import com.google.common.util.concurrent.Futures;
 import com.google.common.util.concurrent.ListenableFuture;
 import com.google.common.util.concurrent.SettableFuture;
+import com.spotify.hype.FluentBackoff;
 import java.io.File;
 import java.io.FileNotFoundException;
 import java.io.IOException;

--- a/hype-gcs/src/main/java/com/spotify/hype/gcs/StagingUtil.java
+++ b/hype-gcs/src/main/java/com/spotify/hype/gcs/StagingUtil.java
@@ -18,7 +18,6 @@
 
 package com.spotify.hype.gcs;
 
-
 import static com.google.cloud.storage.contrib.nio.CloudStorageOptions.withMimeType;
 import static java.nio.file.StandardOpenOption.CREATE_NEW;
 import static java.nio.file.StandardOpenOption.WRITE;

--- a/hype-submitter/src/main/java/com/spotify/hype/runner/KubernetesDockerRunner.java
+++ b/hype-submitter/src/main/java/com/spotify/hype/runner/KubernetesDockerRunner.java
@@ -118,9 +118,8 @@ public class KubernetesDockerRunner implements DockerRunner {
             retrySleeper.sleep(sleep);
           }
         } catch (IOException | InterruptedException ioe) {
-          LOG.error("Failed when trying to sleep: {}", ioe.getMessage(), ioe);
           throw new RuntimeException(
-              String.format("Failed when trying to sleep: {}", ioe.getMessage()), ioe);
+              String.format("Failed to create Kubernetes pod when trying to sleep: %s", ioe.getMessage()), ioe);
         }
       } catch (InterruptedException ie) {
         throw new RuntimeException("Interrupted while blocking", ie);

--- a/hype-submitter/src/main/java/com/spotify/hype/runner/KubernetesDockerRunner.java
+++ b/hype-submitter/src/main/java/com/spotify/hype/runner/KubernetesDockerRunner.java
@@ -28,7 +28,10 @@ import static java.util.stream.Collectors.toMap;
 
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
+import com.google.api.client.util.BackOff;
+import com.google.api.client.util.Sleeper;
 import com.google.common.annotations.VisibleForTesting;
+import com.spotify.hype.FluentBackoff;
 import com.spotify.hype.model.RunEnvironment;
 import com.spotify.hype.model.Secret;
 import com.spotify.hype.model.StagedContinuation;
@@ -81,6 +84,9 @@ public class KubernetesDockerRunner implements DockerRunner {
   private final KubernetesClient client;
   private final VolumeRepository volumeRepository;
 
+  private static final FluentBackoff BACKOFF_FACTORY =
+      FluentBackoff.DEFAULT;
+
   KubernetesDockerRunner(KubernetesClient client, VolumeRepository volumeRepository) {
     this.client = Objects.requireNonNull(client);
     this.volumeRepository = Objects.requireNonNull(volumeRepository);
@@ -88,18 +94,37 @@ public class KubernetesDockerRunner implements DockerRunner {
 
   @Override
   public Optional<URI> run(RunSpec runSpec) {
-    try {
-      final Pod pod = client.pods().create(createPod(runSpec));
-      final String podName = pod.getMetadata().getName();
-      LOG.info("Created pod {}", podName);
+    Sleeper retrySleeper = Sleeper.DEFAULT;
+    BackOff backoff = BACKOFF_FACTORY.backoff();
 
-      Optional<URI> uri = blockUntilComplete(podName);
-      client.pods().withName(podName).delete();
-      return uri;
-    } catch (KubernetesClientException kce) {
-      throw new RuntimeException("Failed to create Kubernetes pod", kce);
-    } catch (InterruptedException e) {
-      throw new RuntimeException("Interrupted while blocking", e);
+    while (true) {
+      try {
+        final Pod pod = client.pods().create(createPod(runSpec));
+        final String podName = pod.getMetadata().getName();
+        LOG.info("Created pod {}", podName);
+
+        Optional<URI> uri = blockUntilComplete(podName);
+        client.pods().withName(podName).delete();
+        return uri;
+      } catch (KubernetesClientException kce) {
+        try {
+          long sleep = backoff.nextBackOffMillis();
+          if (sleep == BackOff.STOP) {
+            // Rethrow last error, to be included as a cause in the catch below.
+            LOG.error("Failed to create Kubernetes pod", kce);
+            throw new KubernetesClientException("Failed to create Kubernetes pod", kce);
+          } else {
+            LOG.warn("Kubernetes creation attempt failed, sleeping before retrying", kce);
+            retrySleeper.sleep(sleep);
+          }
+        } catch (IOException | InterruptedException ioe) {
+          LOG.error("Failed when trying to sleep: {}", ioe.getMessage(), ioe);
+          throw new RuntimeException(
+              String.format("Failed when trying to sleep: {}", ioe.getMessage()), ioe);
+        }
+      } catch (InterruptedException ie) {
+        throw new RuntimeException("Interrupted while blocking", ie);
+      }
     }
   }
 

